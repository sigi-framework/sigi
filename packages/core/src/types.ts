/* eslint-disable @typescript-eslint/no-invalid-void-type */
import { Action } from '@sigi/types'
import { Draft } from 'immer'
import { Observable } from 'rxjs'

import { EffectModule } from './module'

type UnpackEffectPayload<Func> = Func extends (action$: Observable<infer Payload>) => Observable<Action>
  ? Payload
  : never

type UnpackReducerPayload<Func, S> = Func extends (state: S) => S
  ? void
  : Func extends (State: S, payload: infer Payload) => S
    ? Payload
    : never

type UnpackImmerReducerPayload<Func, S> = Func extends (state: Draft<S>) => void
  ? void
  : Func extends (state: Draft<S>, payload: infer Payload) => void
    ? Payload
    : never

type UnpackDefineActionPayload<OB> = OB extends Observable<infer P> ? P : never

type UnpackPayload<F, S> =
  UnpackEffectPayload<F> extends never
    ? UnpackImmerReducerPayload<F, S> extends never
      ? UnpackReducerPayload<F, S> extends never
        ? UnpackDefineActionPayload<F> extends never
          ? never
          : UnpackDefineActionPayload<F>
        : UnpackReducerPayload<F, S>
      : UnpackImmerReducerPayload<F, S>
    : UnpackEffectPayload<F>

export type ActionOfEffectModule<M extends EffectModule<S>, S> = Omit<
  {
    [key in keyof M]: UnpackPayload<M[key], S> extends void ? () => void : (payload: UnpackPayload<M[key], S>) => void
  },
  keyof EffectModule<S>
> & {
  terminate: () => void
  reset: () => void
  noop: () => void
}

export type InstanceActionOfEffectModule<M extends EffectModule<S>, S> = Omit<
  {
    [key in keyof M]: (payload: UnpackPayload<M[key], S>) => Action<UnpackPayload<M[key], S>>
  },
  keyof EffectModule<S>
>

export type RetryActionOfEffectModule<M extends EffectModule<S>, S> = Omit<
  {
    [key in keyof M]: () => Action
  },
  keyof EffectModule<S>
>

export type ActionStreamOfEffectModule<M extends EffectModule<S>, S> = Omit<
  {
    [key in keyof M]: Observable<UnpackPayload<M[key], S>>
  },
  keyof EffectModule<S>
>

export type StateInEffectModule<M extends EffectModule<any>> = M extends EffectModule<infer S> ? S : never
