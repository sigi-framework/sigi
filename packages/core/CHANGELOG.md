# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.13.2](https://github.com/sigi-framework/sigi/compare/v2.13.1...v2.13.2) (2025-06-20)


### Bug Fixes

* **core:** hmr in vite ([23dcbdc](https://github.com/sigi-framework/sigi/commit/23dcbdc0df2b791a5fbbf2450f01de472d54e173))





## [2.13.1](https://github.com/sigi-framework/sigi/compare/v2.13.0...v2.13.1) (2025-06-20)

**Note:** Version bump only for package @sigi/core





# [2.13.0](https://github.com/sigi-framework/sigi/compare/v2.12.2...v2.13.0) (2024-04-22)


### Features

* upgrade dependencies ([fe81b78](https://github.com/sigi-framework/sigi/commit/fe81b785aefabdda40ecc920290c569c4d29a851))





## [2.12.1](https://github.com/sigi-framework/sigi/compare/v2.12.0...v2.12.1) (2021-09-18)

### Bug Fixes

- **react:** state selector may use old closures ([69b19ce](https://github.com/sigi-framework/sigi/commit/69b19ced8f5da2802edf8c1dc793300af448b19e))

## [2.11.3](https://github.com/sigi-framework/sigi/compare/v2.11.2...v2.11.3) (2021-08-30)

**Note:** Version bump only for package @sigi/core

## [2.11.2](https://github.com/sigi-framework/sigi/compare/v2.11.1...v2.11.2) (2021-08-24)

### Bug Fixes

- **core:** skipCount should less equal than the filtered index ([ff38eb6](https://github.com/sigi-framework/sigi/commit/ff38eb6667cc8f0274a893211669925cd3e7f334))

## [2.10.8](https://github.com/sigi-framework/sigi/compare/v2.10.7...v2.10.8) (2021-06-16)

### Bug Fixes

- **ssr:** skipped actions should always be retried on client ([4000e27](https://github.com/sigi-framework/sigi/commit/4000e27bc1ca2ccc3f8e5a8440ce3384b218d8e1))

## [2.10.7](https://github.com/sigi-framework/sigi/compare/v2.10.6...v2.10.7) (2021-06-15)

### Bug Fixes

- **core:** skip from SSR logic ([c3e1ee8](https://github.com/sigi-framework/sigi/commit/c3e1ee8ee97732294c7bf9f3759dd4d7e284e041))

## [2.10.4](https://github.com/sigi-framework/sigi/compare/v2.10.3...v2.10.4) (2021-06-09)

### Bug Fixes

- **core:** compatible with es class property ([3f0e38a](https://github.com/sigi-framework/sigi/commit/3f0e38a8531f72ec62f50552c9dd866945245665))

## [2.10.3](https://github.com/sigi-framework/sigi/compare/v2.10.2...v2.10.3) (2021-06-09)

**Note:** Version bump only for package @sigi/core

## [2.10.2](https://github.com/sigi-framework/sigi/compare/v2.10.1...v2.10.2) (2021-05-26)

**Note:** Version bump only for package @sigi/core

# [2.10.0](https://github.com/sigi-framework/sigi/compare/v2.9.3...v2.10.0) (2021-05-13)

### Features

- **ssr:** force all epic to complete after disposed ([d89786a](https://github.com/sigi-framework/sigi/commit/d89786a4ead1ce84901543274a54f6ea6fa03e9b))

# [2.9.0](https://github.com/sigi-framework/sigi/compare/v2.8.6...v2.9.0) (2021-05-11)

### Bug Fixes

- dependencies list ([4a44cdd](https://github.com/sigi-framework/sigi/commit/4a44cdd65b1fa83ba949f6993c2708d2f6c2b722))

### Features

- **core:** ssr effect now could be marked retry on client ([116df59](https://github.com/sigi-framework/sigi/commit/116df5952c8f45cda94a0d352a5c538b78307ffe))

### Performance Improvements

- **core:** reduce js engine hidden path mismatch ([3224344](https://github.com/sigi-framework/sigi/commit/32243447ba72b7aa957cc52ef7460aaef118e175))
- **react:** fast code path on server ([f7df497](https://github.com/sigi-framework/sigi/commit/f7df49708808f5421f40798151108280ab669359))
- **ssr:** add fast path for ssr di logic ([8cd2f5a](https://github.com/sigi-framework/sigi/commit/8cd2f5a4517fed2f01058cc39d410af63fa53a3b))

## [2.8.6](https://github.com/sigi-framework/sigi/compare/v2.8.5...v2.8.6) (2021-05-06)

**Note:** Version bump only for package @sigi/core

## [2.8.5](https://github.com/sigi-framework/sigi/compare/v2.8.4...v2.8.5) (2021-04-28)

**Note:** Version bump only for package @sigi/core

## [2.8.2](https://github.com/sigi-framework/sigi/compare/v2.8.1...v2.8.2) (2021-03-29)

**Note:** Version bump only for package @sigi/core

# [2.8.0](https://github.com/sigi-framework/sigi/compare/v2.7.0...v2.8.0) (2021-02-22)

**Note:** Version bump only for package @sigi/core

## [2.5.1](https://github.com/sigi-framework/sigi/compare/v2.5.0...v2.5.1) (2021-01-22)

### Bug Fixes

- **core:** setup store immediately after set default state ([d5d2631](https://github.com/sigi-framework/sigi/commit/d5d2631898645a6eb938f7cb8ff5bacac8a9efe6))

## [2.4.5](https://github.com/sigi-framework/sigi/compare/v2.4.4...v2.4.5) (2020-12-14)

### Bug Fixes

- **core:** epic do not follow rxjs control ([e27ab82](https://github.com/sigi-framework/sigi/commit/e27ab824d9b99846654087b840c8b2cccea9af6c))

## [2.4.4](https://github.com/sigi-framework/sigi/compare/v2.4.3...v2.4.4) (2020-11-27)

**Note:** Version bump only for package @sigi/core

## [2.4.3](https://github.com/sigi-framework/sigi/compare/v2.4.2...v2.4.3) (2020-11-22)

**Note:** Version bump only for package @sigi/core

## [2.4.2](https://github.com/sigi-framework/sigi/compare/v2.4.1...v2.4.2) (2020-10-15)

**Note:** Version bump only for package @sigi/core

## [2.4.1](https://github.com/sigi-framework/sigi/compare/v2.4.0...v2.4.1) (2020-08-27)

**Note:** Version bump only for package @sigi/core

# [2.4.0](https://github.com/sigi-framework/sigi/compare/v2.3.1...v2.4.0) (2020-07-28)

### Features

- **core:** get rid of symbols ([6d7a848](https://github.com/sigi-framework/sigi/commit/6d7a84801e4ce2a7be731ad13e26e56cd0970135))
- **react:** expose common actions to dispatcher hooks ([20dd063](https://github.com/sigi-framework/sigi/commit/20dd0631d2501e4b912210fef1f2b69addadb612))

## [2.3.1](https://github.com/sigi-framework/sigi/compare/v2.3.0...v2.3.1) (2020-07-09)

**Note:** Version bump only for package @sigi/core

# [2.3.0](https://github.com/sigi-framework/sigi/compare/v2.2.0...v2.3.0) (2020-07-05)

**Note:** Version bump only for package @sigi/core

# [2.2.0](https://github.com/sigi-framework/sigi/compare/v2.1.0...v2.2.0) (2020-06-28)

### Bug Fixes

- **core:** dispatching action on non-effects module causes dead loop ([1ad1100](https://github.com/sigi-framework/sigi/commit/1ad1100158bd131d7366fef165d4acca30e6739b))

# [2.1.0](https://github.com/sigi-framework/sigi/compare/v2.0.2...v2.1.0) (2020-06-02)

**Note:** Version bump only for package @sigi/core
