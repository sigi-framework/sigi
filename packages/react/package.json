{
  "name": "@sigi/react",
  "version": "2.12.2",
  "description": "React bindings for sigi framework",
  "keywords": [
    "Hooks",
    "RxJS",
    "TypeScript",
    "React",
    "Sigi"
  ],
  "author": "LongYinan <lynweklm@gmail.com>",
  "homepage": "https://sigi.how",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "esm/index.js",
  "browser": "esm/index.browser.js",
  "esnext": "next/index.browser.js",
  "types": "esm/index.d.ts",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "sideEffects": false,
  "files": [
    "dist",
    "esm",
    "next",
    "es2015",
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sigi-framework/sigi.git"
  },
  "bugs": {
    "url": "https://github.com/sigi-framework/sigi/issues"
  },
  "dependencies": {
    "@sigi/core": "^2.12.1",
    "@sigi/di": "^2.11.3",
    "@sigi/ssr": "^2.12.1",
    "@sigi/types": "^2.11.3",
    "tslib": "^2.6.1"
  },
  "peerDependencies": {
    "react": ">=18",
    "rxjs": ">=7",
    "use-sync-external-store": ">=1"
  },
  "devDependencies": {
    "@types/use-sync-external-store": "^0.0.6",
    "use-sync-external-store": "^1.2.0"
  }
}
