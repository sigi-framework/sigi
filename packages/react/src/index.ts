import { EffectModule, ActionOfEffectModule } from '@sigi/core'
import { ConstructorOf } from '@sigi/types'

import { useServerInstance } from './injectable-context'

export type StateSelector<S, U> = {
  (state: S): U
}

export type StateSelectorConfig<S, U> = {
  selector: StateSelector<S, U>
}

const SERVER_DISPATCHERS = new Proxy(Object.create(null), {
  apply() {
    if (process.env.NODE_ENV !== 'production') {
      console.warn(new Error('Dispatch while calling react-dom/server functions is forbidden'))
    }
  },
})

export function useDispatchers<M extends EffectModule<S>, S = any>(_A: ConstructorOf<M>): ActionOfEffectModule<M, S> {
  return SERVER_DISPATCHERS
}

export function useModuleState<M extends EffectModule<any>>(
  A: ConstructorOf<M>,
): M extends EffectModule<infer State> ? State : never

export function useModuleState<M extends EffectModule<any>, U>(
  A: ConstructorOf<M>,
  config: M extends EffectModule<infer State> ? StateSelectorConfig<State, U> : never,
): M extends EffectModule<infer State>
  ? (typeof config)['selector'] extends StateSelector<State, infer NewState>
    ? NewState
    : never
  : never

export function useModuleState<M extends EffectModule<any>, U>(
  A: ConstructorOf<M>,
  config?: M extends EffectModule<infer S> ? StateSelectorConfig<S, U> : never,
) {
  const { store } = useServerInstance(A)
  return typeof config?.selector === 'function' ? config.selector(store.state) : store.state
}

export function useModule<M extends EffectModule<any>>(
  A: ConstructorOf<M>,
): M extends EffectModule<infer State> ? [State, ActionOfEffectModule<M, State>] : never

export function useModule<M extends EffectModule<any>, U>(
  A: ConstructorOf<M>,
  config: M extends EffectModule<infer State> ? StateSelectorConfig<State, U> : never,
): M extends EffectModule<infer State>
  ? (typeof config)['selector'] extends StateSelector<State, infer NewState>
    ? [NewState, ActionOfEffectModule<M, State>]
    : never
  : never

export function useModule<M extends EffectModule<S>, U, S>(A: ConstructorOf<M>, config?: StateSelectorConfig<S, U>) {
  const effectModule = useServerInstance(A)
  const { store } = effectModule
  const appState = typeof config?.selector === 'function' ? config.selector(store.state) : store.state

  return [appState, SERVER_DISPATCHERS]
}

export { SSRContext } from './ssr-context'
export * from './injectable-context'
