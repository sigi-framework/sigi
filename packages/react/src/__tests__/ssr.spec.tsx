/* eslint-disable sonarjs/no-identical-functions */
import 'reflect-metadata'

import { GLOBAL_KEY_SYMBOL, EffectModule, ImmerReducer, Module, Effect, Reducer, RETRY_KEY_SYMBOL } from '@sigi/core'
import { Injectable, Injector } from '@sigi/di'
import { emitSSREffects, match } from '@sigi/ssr'
import { Action } from '@sigi/types'
import { Draft } from 'immer'
import { useEffect } from 'react'
import { renderToString } from 'react-dom/server'
import { create, act } from 'react-test-renderer'
import { Observable, of, timer } from 'rxjs'
import { endWith, switchMap, map, mergeMap, withLatestFrom } from 'rxjs/operators'

import { SSRContext, useModule } from '../index.browser'

interface CountState {
  count: number
  name: string
}

interface TipState {
  tip: string
}

@Injectable()
class Service {
  getName() {
    return of('client service')
  }
}

@Module('ServiceModule')
class ServiceModule extends EffectModule<CountState> {
  readonly defaultState = { count: 0, name: '' }

  constructor(public readonly service: Service) {
    super()
  }

  @ImmerReducer()
  setName(state: Draft<CountState>, name: string) {
    state.name = name
  }

  @Effect({
    ssr: true,
  })
  setNameEffect(payload$: Observable<void>): Observable<Action> {
    return payload$.pipe(
      switchMap(() =>
        this.service.getName().pipe(
          map((name) => this.getActions().setName(name)),
          endWith(this.terminate()),
        ),
      ),
    )
  }

  @Effect({
    payloadGetter: (ctx: { failure: boolean }, skip) => {
      if (!ctx.failure) {
        return skip
      }
      return 1
    },
  })
  setNameWithFailure(payload$: Observable<number | undefined>): Observable<Action> {
    return payload$.pipe(
      mergeMap((p) => {
        if (p) {
          return of(this.retryOnClient().setNameWithFailure(), this.terminate())
        }
        return of(this.getActions().setName('From retry'))
      }),
    )
  }
}

@Module('CountModel')
class CountModel extends EffectModule<CountState> {
  defaultState = { count: 0, name: '' }

  constructor(public readonly serviceModule: ServiceModule) {
    super()
  }

  @ImmerReducer()
  setCount(state: Draft<CountState>, count: number) {
    state.count = count
  }

  @ImmerReducer()
  setName(state: Draft<CountState>, name: string) {
    state.name = name
  }

  @Effect({
    ssr: true,
  })
  getCount(payload$: Observable<void>): Observable<Action> {
    return payload$.pipe(
      mergeMap(() =>
        timer(20).pipe(
          map(() => this.getActions().setCount(1)),
          endWith(this.terminate()),
        ),
      ),
    )
  }

  @Effect({
    payloadGetter: (ctx: { url: string }, skip) => ctx.url || skip,
    skipFirstClientDispatch: false,
  })
  skippedEffect(payload$: Observable<string>): Observable<Action> {
    return payload$.pipe(
      switchMap((name) =>
        timer(20).pipe(
          map(() => this.getActions().setName(name)),
          endWith(this.terminate()),
        ),
      ),
    )
  }

  @Effect({
    payloadGetter: (ctx: { name: string }, skip) => ctx.name || skip,
  })
  emitOtherEffect(payload$: Observable<string>): Observable<Action> {
    return payload$.pipe(
      switchMap((name) =>
        timer(20).pipe(
          map(() => this.serviceModule.getActions().setName(name)),
          endWith(this.terminate()),
        ),
      ),
    )
  }
}

interface InfinityWaitModelState {
  count: number
}

@Module('InfinityWaitModel')
class InfinityWaitModel extends EffectModule<InfinityWaitModelState> {
  defaultState = { count: 0 }

  @ImmerReducer()
  setCount(state: Draft<InfinityWaitModelState>, count: number) {
    state.count = count
  }

  @Effect({
    ssr: true,
  })
  infinityWait(payload$: Observable<void>): Observable<Action> {
    return payload$.pipe(map(() => this.getActions().setCount(1)))
  }
}

@Module('TipModel')
class TipModel extends EffectModule<TipState> {
  defaultState = { tip: '' }

  @ImmerReducer()
  setTip(state: Draft<TipState>, tip: string) {
    state.tip = tip
  }

  @Effect({ ssr: true })
  getTip(payload$: Observable<void>): Observable<Action> {
    return payload$.pipe(
      mergeMap(() =>
        timer(1).pipe(
          map(() => this.getActions().setTip('tip')),
          endWith(this.terminate()),
        ),
      ),
    )
  }
}

const Component = () => {
  const [state, actions] = useModule(CountModel)
  useEffect(() => {
    actions.setName('new name')
  }, [actions])

  return (
    <>
      <span>{state.count}</span>
    </>
  )
}

const ComponentWithSelector = () => {
  const [state, actions] = useModule(CountModel, {
    selector: (s) => ({
      count: s.count + 1,
    }),
    dependencies: [],
  })
  useEffect(() => {
    actions.setName('new name')
  }, [actions])

  return (
    <>
      <span>{state.count}</span>
    </>
  )
}

const MODULES = [CountModel, TipModel, ServiceModule, Service]

describe('SSR specs:', () => {
  it('should throw if module name not given', () => {
    function generateException() {
      // @ts-expect-error
      @Module()
      class ErrorModel extends EffectModule<any> {
        defaultState = {}
      }

      return ErrorModel
    }

    expect(generateException).toThrow()
  })

  it('should pass valid module name', () => {
    @Module('1')
    class Model extends EffectModule<any> {
      defaultState = {}
    }

    @Module('2')
    class Model2 extends EffectModule<any> {
      defaultState = {}
    }

    function generateException1() {
      @Module('1')
      class ErrorModel1 extends EffectModule<any> {
        defaultState = {}
      }

      return ErrorModel1
    }

    function generateException2() {
      @Module('1')
      class ErrorModel2 extends EffectModule<any> {
        defaultState = {}
      }

      return { ErrorModel2 }
    }

    // eslint-disable-next-line sonarjs/no-identical-functions
    function generateException3() {
      // @ts-expect-error
      @Module()
      class ErrorModel extends EffectModule<any> {
        defaultState = {}
      }

      return ErrorModel
    }

    expect(Model).not.toBe(undefined)
    expect(Model2).not.toBe(undefined)
    expect(generateException1).toThrow()
    expect(generateException2).toThrow()
    expect(generateException3).toThrow()
  })

  it('should run ssr effects', async () => {
    const state = await emitSSREffects({ url: 'name' } as any, [CountModel], {
      providers: MODULES,
    }).pendingState
    const moduleState = state['dataToPersist']['CountModel']
    expect(moduleState).not.toBe(undefined)
    expect(moduleState.count).toBe(1)
    expect(moduleState.name).toBe('name')
    expect(state['dataToPersist']).toMatchSnapshot()
  })

  it('should skip effect if it returns SKIP_SYMBOL', async () => {
    const state = await emitSSREffects({}, [CountModel], {
      providers: MODULES,
    }).pendingState
    const moduleState = state['dataToPersist']['CountModel']

    expect(moduleState.name).toBe('')
    expect(state['dataToPersist']).toMatchSnapshot()
  })

  it("should return right state if emit other module's effect", async () => {
    const state = await emitSSREffects({ name: 'foo' }, [CountModel], {
      providers: [],
    }).pendingState
    const moduleState = state['dataToPersist']['ServiceModule']
    expect(moduleState.name).toBe('foo')
    expect(state['dataToPersist']).toMatchSnapshot()
  })

  it('should return right state in hooks', async () => {
    const req = {}
    const { pendingState, injector } = emitSSREffects(req, [CountModel], {
      providers: MODULES,
    })
    await pendingState
    const html = renderToString(
      <SSRContext value={injector}>
        <Component />
      </SSRContext>,
    )
    expect(html).toContain('<span>1</span>')
    expect(html).toMatchSnapshot()
  })

  it('should restore state from global', () => {
    global[GLOBAL_KEY_SYMBOL] = {
      CountModel: {
        count: 101,
        name: '',
      },
    }
    const testRenderer = create(<Component />)
    act(() => {
      testRenderer.update(<Component />)
    })
    expect(testRenderer.root.findByType('span').children[0]).toBe('101')

    delete global[GLOBAL_KEY_SYMBOL]
    testRenderer.unmount()
  })

  it('should restore state from global #with selector', () => {
    global[GLOBAL_KEY_SYMBOL] = {
      CountModel: {
        count: 10,
        name: '',
      },
    }
    const testRenderer = create(
      <SSRContext value={new Injector().addProviders(MODULES)}>
        <ComponentWithSelector />
      </SSRContext>,
    )
    expect(testRenderer.root.findByType('span').children[0]).toBe('11')
    delete global[GLOBAL_KEY_SYMBOL]
    testRenderer.unmount()
  })

  it('should not restore state from global if state is null', () => {
    global[GLOBAL_KEY_SYMBOL] = {
      OtherModule: {
        count: 10,
        name: '',
      },
    }
    const injector = new Injector().addProviders(MODULES)
    const testRenderer = create(
      <SSRContext value={injector}>
        <ComponentWithSelector />
      </SSRContext>,
    )
    act(() => {
      testRenderer.update(
        <SSRContext value={injector}>
          <ComponentWithSelector />
        </SSRContext>,
      )
    })
    expect(testRenderer.root.findByType('span').children[0]).toBe('1')

    delete global[GLOBAL_KEY_SYMBOL]
    testRenderer.unmount()
  })

  it('should restore and skip first action on client side', () => {
    const Component = () => {
      const [state, actions] = useModule(CountModel)
      useEffect(() => {
        actions.getCount()
      }, [actions])

      return (
        <>
          <span>{state.count}</span>
        </>
      )
    }

    global[GLOBAL_KEY_SYMBOL] = {
      CountModel: {
        count: 2,
        name: '',
      },
    }

    const injector = new Injector().addProviders(MODULES)

    const testRenderer = create(
      <SSRContext value={injector}>
        <Component />
      </SSRContext>,
    )

    act(() => {
      testRenderer.update(
        <SSRContext value={injector}>
          <Component />
        </SSRContext>,
      )
    })
    expect(testRenderer.root.findByType('span').children[0]).toBe('2')

    delete global[GLOBAL_KEY_SYMBOL]
    testRenderer.unmount()
  })

  it('should retry action if needed on client side', () => {
    const Component = () => {
      const [state, dispatcher] = useModule(ServiceModule)
      useEffect(() => {
        dispatcher.setNameWithFailure(void 0)
      }, [dispatcher])

      return (
        <>
          <span>{state.name}</span>
        </>
      )
    }

    global[RETRY_KEY_SYMBOL] = {
      ServiceModule: ['setNameWithFailure'],
    }

    const injector = new Injector().addProviders(MODULES)

    const testRenderer = create(
      <SSRContext value={injector}>
        <Component />
      </SSRContext>,
    )

    // eslint-disable-next-line sonarjs/no-identical-functions
    act(() => {
      testRenderer.update(
        <SSRContext value={injector}>
          <Component />
        </SSRContext>,
      )
    })

    expect(testRenderer.root.findByType('span').children[0]).toBe('From retry')

    delete global[GLOBAL_KEY_SYMBOL]
    testRenderer.unmount()
  })

  it('should timeout', async () => {
    const req = {}
    return emitSSREffects(req, [CountModel], { providers: MODULES, timeout: 0 }).pendingState.catch((e: Error) => {
      expect(e.message).toBe('Terminate timeout')
    })
  })

  // Do not use `Sinon.fakeTimers` here
  // It would cause `UnhandledPromiseRejection`
  it('should timeout #2', async () => {
    const req = {}
    const { pendingState } = emitSSREffects(req, [InfinityWaitModel, TipModel], {
      timeout: 2 / 1000,
    })
    await pendingState.catch((e) => {
      expect(e.message).toBe('Terminate timeout')
    })
  })

  it('should resolve empty object if no modules provided', async () => {
    const req = {}
    const state = await emitSSREffects(req, [], { providers: MODULES }).pendingState
    expect(state['dataToPersist']).toStrictEqual({})
  })

  it('should do nothing if Module contains no SSREffects', async () => {
    const req = {}
    @Module('WithoutSSR')
    class WithoutSSRModule extends EffectModule<{ count: number }> {
      defaultState = {
        count: 0,
      }

      @ImmerReducer()
      set(state: Draft<{ count: number }>, payload: number) {
        state.count = payload
      }

      @Effect()
      addOne(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(
          withLatestFrom(this.state$),
          map(([, { count }]) => this.getActions().set(count + 1)),
        )
      }
    }
    const state = await emitSSREffects(req, [WithoutSSRModule], { providers: MODULES }).pendingState
    expect(state['dataToPersist']).toStrictEqual({
      WithoutSSR: {
        count: 0,
      },
    })
  })

  it('should throw error if runEffects error', async () => {
    const req = {}
    const error = new TypeError('whatever')
    @Module('ErrorModule')
    class SSRErrorModule extends EffectModule<{ count: number }> {
      defaultState = {
        count: 0,
      }

      @Effect({ ssr: true })
      addOne(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(
          withLatestFrom(this.state$),
          map(() => {
            throw error
          }),
        )
      }
    }
    try {
      await emitSSREffects(req, [SSRErrorModule], { providers: MODULES }).pendingState
      throw new TypeError('Unreachable code path')
    } catch (e) {
      expect(e).toBe(error)
    }
  })

  it('should throw error if reducer throw', async () => {
    const req = {}
    const error = new TypeError('whatever')
    @Module('ErrorReducerModule')
    class SSRErrorModule extends EffectModule<{ count: number }> {
      defaultState = {
        count: 0,
      }

      @Reducer()
      set() {
        throw error
      }

      @Effect({ ssr: true })
      addOne(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(
          withLatestFrom(this.state$),
          map(() => this.getActions().set()),
        )
      }
    }
    try {
      await emitSSREffects(req, [SSRErrorModule], { providers: MODULES }).pendingState
      throw new TypeError('Unreachable code path')
    } catch (e) {
      expect(e).toBe(error)
    }
  })

  it('should throw error if payloadGetter throw', async () => {
    const req = {}
    const error = new TypeError('whatever')
    @Module('ErrorMiddlewareModule')
    class SSRErrorModule extends EffectModule<{ count: number }> {
      defaultState = {
        count: 0,
      }

      @Reducer()
      set(state: { count: number }, payload: number) {
        return { ...state, count: payload }
      }

      @Effect({
        payloadGetter: () => {
          throw error
        },
      })
      addOne(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(
          withLatestFrom(this.state$),
          map(([, state]) => this.getActions().set(state.count + 1)),
        )
      }
    }
    try {
      await emitSSREffects(req, [SSRErrorModule], { providers: MODULES }).pendingState
      throw new TypeError('Unreachable code path')
    } catch (e) {
      expect(e).toBe(error)
    }
  })

  it('should replace injector if providers provided', async () => {
    const req = {}
    const state = await emitSSREffects(req, [ServiceModule], { providers: MODULES }).pendingState
    expect(state['dataToPersist'].ServiceModule.name).toBe('client service')
    const state2 = await emitSSREffects(req, [ServiceModule], {
      providers: [...MODULES, { provide: Service, useValue: { getName: () => of('server service') } }],
    }).pendingState
    expect(state2['dataToPersist'].ServiceModule.name).toBe('server service')
  })

  it('should persist states which mutated by the other modules', async () => {
    @Module('InnerStateModule')
    class StateModel extends EffectModule<{ count: number }> {
      defaultState = { count: 0 }

      @ImmerReducer()
      setCount(state: Draft<{ count: number }>, count: number) {
        state.count = count
      }
    }

    @Module('InnerCountModel')
    // eslint-disable-next-line @typescript-eslint/ban-types
    class CountModel extends EffectModule<{}> {
      defaultState = {}

      constructor(private readonly stateModule: StateModel) {
        super()
      }

      @Effect({
        ssr: true,
      })
      getCount(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(
          mergeMap(() =>
            timer(20).pipe(
              map(() => this.stateModule.getActions().setCount(1)),
              endWith(this.terminate()),
            ),
          ),
        )
      }
    }

    const req = {}
    const state = await emitSSREffects(req, [CountModel, StateModel]).pendingState

    expect(state['dataToPersist']).toEqual({
      InnerCountModel: {},
      InnerStateModule: {
        count: 1,
      },
    })
  })

  it('should persist actions to retry if needed', async () => {
    const req = { failure: true }
    const state = await emitSSREffects(req, [ServiceModule], { providers: MODULES }).pendingState
    expect(state['actionsToRetry']).toEqual({ ServiceModule: ['setNameWithFailure'] })
  })

  it('should be able to persist actions to retry on the other module, and the skipped actions', async () => {
    @Module('InnerServiceModule')
    // eslint-disable-next-line @typescript-eslint/ban-types
    class InnerServiceModule extends EffectModule<CountState> {
      readonly defaultState = { count: 0, name: '' }

      constructor(public readonly service: Service) {
        super()
      }

      @ImmerReducer()
      setName(state: Draft<CountState>, name: string) {
        state.name = name
      }

      @Effect()
      setNameWithFailure(payload$: Observable<number | undefined>): Observable<Action> {
        return payload$.pipe(mergeMap(() => of(this.retryOnClient().setNameWithFailure(), this.terminate())))
      }
    }
    @Module('InnerCountModel2')
    // eslint-disable-next-line @typescript-eslint/ban-types
    class InnerCountModule2 extends EffectModule<{}> {
      defaultState = {}

      constructor(private readonly serviceModule: InnerServiceModule) {
        super()
      }

      @Effect({
        ssr: true,
      })
      setName(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(mergeMap(() => of(this.serviceModule.getActions().setNameWithFailure(1))))
      }

      @Effect({
        payloadGetter(_: any, skipAction) {
          return skipAction
        },
      })
      skippedSetName(payload$: Observable<void>): Observable<Action> {
        return payload$.pipe(map(() => this.serviceModule.getActions().setName('skipped')))
      }
    }

    const req = {}
    const state = await emitSSREffects(req, [InnerServiceModule, InnerCountModule2], { providers: MODULES })
      .pendingState

    expect(state['actionsToRetry']).toEqual({
      InnerCountModel2: ['skippedSetName'],
      InnerServiceModule: ['setNameWithFailure'],
    })
  })

  it('should support match fn', async () => {
    @Module('InnerServiceModule2')
    class InnerServiceModule2 extends EffectModule<CountState> {
      readonly defaultState = { count: 0, name: '' }

      @ImmerReducer()
      setName(state: Draft<CountState>, name: string) {
        state.name = name
      }

      @Effect({
        payloadGetter: match(
          ['/users/:id'],
          (ctx: any) => ctx.request.path,
        )((ctx) => {
          return ctx.request.path.length
        }),
      })
      setNameWithFailure(payload$: Observable<number>): Observable<Action> {
        return payload$.pipe(mergeMap((l) => of(this.getActions().setName(`length: ${l}`), this.terminate())))
      }
    }

    const req = {
      request: {
        path: '/users/linus',
      },
    }
    const state = await emitSSREffects(req, [InnerServiceModule2], { providers: [InnerServiceModule2] }).pendingState

    expect(state['dataToPersist']).toEqual({
      InnerServiceModule2: {
        count: 0,
        name: `length: ${req.request.path.length}`,
      },
    })
  })
})
