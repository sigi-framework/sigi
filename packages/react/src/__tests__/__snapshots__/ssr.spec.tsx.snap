// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR specs: should return right state if emit other module's effect 1`] = `
Object {
  "CountModel": Object {
    "count": 1,
    "name": "",
  },
  "ServiceModule": Object {
    "count": 0,
    "name": "foo",
  },
}
`;

exports[`SSR specs: should return right state in hooks 1`] = `"<span>1</span>"`;

exports[`SSR specs: should run ssr effects 1`] = `
Object {
  "CountModel": Object {
    "count": 1,
    "name": "name",
  },
}
`;

exports[`SSR specs: should skip effect if it returns SKIP_SYMBOL 1`] = `
Object {
  "CountModel": Object {
    "count": 1,
    "name": "",
  },
}
`;
